### DISCUSSION TITLE
How to get annotations of an annotation?

### DISCUSSION THREAD

**Discussion by evilpan** [soot-oss/SootUp#616] [Q&A] [answered]
For example, some class is annotated by `RestController`:
```java
@RestController
@RequestMapping({"/rest/"})
public class DemoRestController {

   @GetMapping({"/foo"})
   public String foo() throws Exception {
     return "foo";
   }
}
```

which `@RestController` is annotated by other annotations:
```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Controller
@ResponseBody
public @interface RestController {

	@AliasFor(annotation = Controller.class)
	String value() default "";

}
```

I can get AnnotationUsage of a JavaSootClass via:
```java
cls.getAnnotations(Optional.of(view));
```

But how to get list of annotation of each annotation usage of above results? i.e. `@Controller`, `@ResponseBody`, etc.

**Comment by kadirayk** [comment]
Did you try getting the SootClass of RestController and then calling .getAnnotations on it?

**Reply by JonasKlauke** [reply, marked as answer]
Yeah this works.
For example if au is your AnnotationUsage contained in the resulting list of  cls.getAnnotations(Optional.of(view)); then you get the AnnotationType by calling `au.getAnnotation()`. The AnnotationType can be used to get the SootClass by `view.getClass(au.getAnnotation())` and on this you can call getAnnotations() again `view.getClass(au.getAnnotation()).ifPresent(javaSootClass -> javaSootClass.getAnnotations(Optional.of(view)));`

**Comment by evilpan** [comment]
Thanks!
